File: java/plugins/com/voxgig/sekreto/plugins/Proc.java
// Running a child process to completion, shared by the two plugins that
// read a secret through somebody else's CLI - boru and secretspec.
//
// In the core of no port: a kind that spawns a process is a plugin.

package com.voxgig.sekreto.plugins;

import com.voxgig.sekreto.Sekreto.SekretoError;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;

public final class Proc {

  private Proc() {}

  /** What a finished child process left behind. */
  public static final class Ran {
    final String out;
    final String why;
    final int status;

    Ran(String out, String why, int status) {
      this.out = out;
      this.why = why;
      this.status = status;
    }
  }

  /**
   * Run a child to completion and collect both its streams.
   *
   * <p>The two streams are drained CONCURRENTLY. Reading stdout to EOF and
   * only then reading stderr deadlocks the moment the child writes more than
   * one pipe buffer (64 KiB on Linux) to stderr: the parent is blocked
   * waiting for stdout, the child is blocked waiting for room on stderr, and
   * neither can move. Nothing in this library sets a timeout, so that hang is
   * permanent - `get()` simply never returns. secretspec's diagnostics are
   * box-drawn and reach that size easily.
   *
   * <p>The child's stdin is closed rather than left open on a pipe nobody
   * writes to, so a CLI that reads it - one prompting for a passphrase when
   * its environment variable is absent - sees EOF and gives up instead of
   * waiting forever.
   */
  public static Ran runcmd(ProcessBuilder builder, String command) {
    try {
      Process process = builder.start();

      process.getOutputStream().close();

      ByteArrayOutputStream errbuf = new ByteArrayOutputStream();
      Thread drain =
          new Thread(
              () -> {
                try {
                  process.getErrorStream().transferTo(errbuf);
                } catch (IOException err) {
                  // The child went away mid-write; waitFor reports how.
                }
              });
      drain.setDaemon(true);
      drain.start();

      String out = new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
      int status = process.waitFor();
      drain.join();

      return new Ran(out, new String(errbuf.toByteArray(), StandardCharsets.UTF_8).trim(), status);
    } catch (IOException err) {
      throw new SekretoError("sekreto: cannot run " + command + ": " + err.getMessage());
    } catch (InterruptedException err) {
      Thread.currentThread().interrupt();
      throw new SekretoError("sekreto: interrupted running " + command);
    }
  }
}

File: clojure/plugins/voxgig/sekreto/plugins/proc.clj
;; Running a child to completion, for the two kinds that read a secret
;; through somebody else's CLI - boru and secretspec.
;;
;; OUTSIDE THE CORE, because spawning a process is one of the three things
;; that make a kind a plugin. A chain of built-ins never loads this
;; namespace and so never reaches `ProcessBuilder`
;; (docs/design/plugin-providers.md).
;;
;; A port of java/plugins/com/voxgig/sekreto/plugins/Proc.java, which
;; carries the same two-stream rule.

(ns voxgig.sekreto.plugins.proc
  (:require [clojure.string :as string]
            [voxgig.sekreto.core :as core])
  (:import [java.io ByteArrayOutputStream IOException]
           [java.nio.charset StandardCharsets]))

(defn runcmd
  "Run a child to completion and collect both its streams.

  The two streams are drained CONCURRENTLY. Reading stdout to EOF and only
  then reading stderr deadlocks the moment the child writes more than one
  pipe buffer (64 KiB on Linux) to stderr: the parent is blocked waiting for
  stdout, the child is blocked waiting for room on stderr, and neither can
  move. Nothing in this library sets a timeout, so that hang is permanent -
  `get` simply never returns. secretspec's diagnostics are box-drawn and
  reach that size easily.

  The child's stdin is closed rather than left open on a pipe nobody writes
  to, so a CLI that reads it - one prompting for a passphrase when its
  environment variable is absent - sees EOF and gives up instead of waiting
  forever."
  [^ProcessBuilder builder command]
  (try
    (let [process (.start builder)]
      (.close (.getOutputStream process))

      (let [errbuf (ByteArrayOutputStream.)
            pump (fn []
                   (try
                     (.transferTo (.getErrorStream process) errbuf)
                     ;; The child went away mid-write; waitFor reports how.
                     (catch IOException _ nil)))
            drain (doto (Thread. ^Runnable pump) (.setDaemon true) (.start))
            out (String. (.readAllBytes (.getInputStream process)) StandardCharsets/UTF_8)
            status (.waitFor process)]
        (.join drain)
        {:out out
         :why (string/trim (String. (.toByteArray errbuf) StandardCharsets/UTF_8))
         :status status}))
    (catch IOException err
      (throw (core/sekretoerror (str "sekreto: cannot run " command ": " (.getMessage err)))))
    (catch InterruptedException _
      (.interrupt (Thread/currentThread))
      (throw (core/sekretoerror (str "sekreto: interrupted running " command))))))

(defn command
  "A ProcessBuilder for an argument vector."
  ^ProcessBuilder [args]
  (ProcessBuilder. ^"[Ljava.lang.String;" (into-array String args)))
